### PR TITLE
tests: Rework tests generators

### PR DESCRIPTION
**Pull request by aarnq**
> [!warning]
> **This is a public repository, ensure not to disclose:**
>
> - [x] personal data beyond what is necessary for interacting with this pull request, nor
> - [x] business confidential information, such as customer names.

### What kind of PR is this?

**Required**: Mark one of the following that is applicable:

- [ ] kind/feature       
- [x] kind/improvement   
- [ ] kind/deprecation   
- [ ] kind/documentation 
- [ ] kind/clean-up      
- [ ] kind/bug           
- [ ] kind/other         

*Optional*: Mark one or more of the following that are applicable:

> [!important]
> Breaking changes should be marked `kind/admin-change` or `kind/dev-change` depending on type
> Critical security fixes should be marked with `kind/security`

- [ ] kind/admin-change 
- [ ] kind/dev-change   
- [ ] kind/security     
- [ ] [kind/adr]()      

### What does this PR do / why do we need this PR?

This reworks the generated bats tests to use go-templates to make the process of writing them be more natural.

This also makes generation part of the pre-commit hook so that one can see the impact of changes.

#### Information to reviewers

Do let me know what you think about this, I see it the template process as an improvement, but it will add a lot of noise when one makes changes as it like here adds ~70 loc for ~20 targets which yields 1400 loc in total. 

#### Checklist

- [x] Proper commit message prefix on all commits
  
- Change checks:
  - [x] The change is transparent
  - [ ] The change is disruptive
  - [x] The change requires no migration steps
  - [ ] The change requires migration steps
  - [ ] The change upgrades CRDs
- Metrics checks:
  - [ ] The metrics are still exposed and present in Grafana after the change
  - [ ] The metrics names didn't change (Grafana dashboards and Prometheus alerts are not affected)
  - [ ] The metrics names did change (Grafana dashboards and Prometheus alerts were fixed)
- Logs checks:
  - [ ] The logs do not show any errors after the change
- Pod Security Policy checks:
  - [ ] Any changed pod is covered by Pod Security Admission
  - [ ] Any changed pod is covered by Gatekeeper Pod Security Policies
  - [ ] The change does not cause any pods to be blocked by Pod Security Admission or Policies
- Network Policy checks:
  - [ ] Any changed pod is covered by Network Policies
  - [ ] The change does not cause any dropped packets in the `NetworkPolicy Dashboard`
- Audit checks:
  - [ ] The change does not cause any unnecessary Kubernetes audit events
  - [ ] The change requires changes to Kubernetes audit policy
- Falco checks:
  - [ ] The change does not cause any alerts to be generated by Falco
- Bug checks:
  - [ ] The bug fix is covered by regression tests
- Config checks:
  - [ ] The schema was updated

### CODE REVIEW

**Comment by simonklb** [comment]
I like it! The fact that you can see the results from your code in the PR is good IMO.

Could you maybe make the generating a Makefile target so that you can do it manually instead of having to commit to do it? If I was writing tests I would like to generate it and diff it in a short loop while developing.

**Comment by aarnq** [comment]
> I like it! The fact that you can see the results from your code in the PR is good IMO.
> 
> Could you maybe make the generating a Makefile target so that you can do it manually instead of having to commit to do it? If I was writing tests I would like to generate it and diff it in a short loop while developing.

Yes, I'd want to make it so that it works as the previous model did so you only have to regenerate them if they changed and I also want to better package everything into a container so the generation can happen through it instead.

This was more of a stop gap to ensure that the formatting works against other hooks.

**Comment by aarnq** [comment]
I think this will be it for this PR, I'll leave the `pods-ready.bats.yaml` for now since it will need more rethinking, and it is not used in the same capacity since we've yet to use it via `./bin/ck8s test`.

I'll also investigate the possibility to use the container more effectively when running tests, but that will be introduced in a different PR as well.

File: tests/common/gen.bash [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+  file="$(readlink -f "$1")"
+
+  local -a input
+
+  readarray -t input < "${file}"
+
+  local describe test
+
+  local -a its
```
Comment (lunkan93) [inline tests/common/gen.bash:20]: Might be a dumb question but what is the acronym here? :sweat_smile:

Comment (aarnq) [reply tests/common/gen.bash:20]: It collects the `it`s declared in the `.cy.js` files, so it is not really an acronym :sweat_smile:

Comment (lunkan93) [reply tests/common/gen.bash:20]: Aaah now it all makes sense, thanks :+1:

**Review by lunkan93 — APPROVED** [review]
I think this looks good :+1: I personally don't mind the noise aspect too much

**Comment by robinAwallace** [comment]
It looks good to me, just one observation :slightly_smiling_face:  I ran the pre-commit hooks just to check how fast the templating was and I got a un-trackted file `tests/unit/bin/init.gen.bats`. Should that be part of this PR? :eyes:

**Comment by aarnq** [comment]
> It looks good to me, just one observation 🙂 I ran the pre-commit hooks just to check how fast the templating was and I got a un-trackted file `tests/unit/bin/init.gen.bats`. Should that be part of this PR? 👀

I think that is left over from the old method, because no file should template into `init.gen.bats`, that one has been replaced by the tests in `tests/unit/bin/init/`.

You can run `make clean-gen` from `tests/` and then `make gen` and it should be up to date without untracked files.

**Review by robinAwallace — APPROVED** [review]

**Pull request merged**